pochivm/api_base.h

```diff
@@ -117,111 +117,223 @@ Value<T>::operator Value<U>() const
+Value<typename AstTypeHelper::ArithReturnType<T, U>::type> operator/(const Value<T>& lhs, const Value<U>& rhs)
 {
-    return Value<T>(new AstArithmeticExpr(AstArithmeticExprType::DIV, lhs.__pochivm_value_ptr, rhs.__pochivm_value_ptr));
+    return DoArith<AstTypeHelper::BinaryOps::DIV>(lhs, rhs, AstArithmeticExprType::DIV);
 }
 
-template<typename T, typename = std::enable_if_t<
-             AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::MODULO>::value> >
-Value<T> operator%(const Value<T>& lhs, const Value<T>& rhs)
+template <typename T, typename U,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<U, AstTypeHelper::BinaryOps::MODULO>::value>,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::MODULO>::value> >
+Value<typename AstTypeHelper::ArithReturnType<T, U>::type> operator%(const Value<T>& lhs, const Value<U>& rhs)
 {
-    return Value<T>(new AstArithmeticExpr(AstArithmeticExprType::MOD, lhs.__pochivm_value_ptr, rhs.__pochivm_value_ptr));
+    return DoArith<AstTypeHelper::BinaryOps::MODULO>(lhs, rhs, AstArithmeticExprType::MOD);
 }
 
 // Convenience overloading: arithmetic operation with literal
 //
-template<typename T, typename = std::enable_if_t<
-             AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::ADD>::value> >
-Value<T> operator+(const Value<T>& lhs, T rhs)
+
+// Do arithmetic of form runtime value OP literal
+//
+template <AstTypeHelper::BinaryOps Binop, typename T, typename U,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<T, Binop>::value>,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<U, Binop>::value> >
+Value<typename AstTypeHelper::ArithReturnType<T, U>::type> DoArith(const Value<T>& lhs, U rhs, const AstArithmeticExprType expr_type)
+{
+    using ReturnType = typename AstTypeHelper::ArithReturnType<T, U>::type;
+    static_assert(std::is_signed<T>::value == std::is_signed<U>::value ||
+                      std::is_floating_point<ReturnType>::value,
+                  "cannot do arithmetic with two values of different signedness");
+    if constexpr (!std::is_same<T, ReturnType>::value)
+    {
+        static_assert(std::is_same<ReturnType, U>::value, "internal bug: rhs type is not the same as the computed common type");
+        return Value<ReturnType>(new AstArithmeticExpr(expr_type,
+                                                       StaticCast<ReturnType>(lhs).__pochivm_value_ptr,
+                                                       new AstLiteralExpr(TypeId::Get<ReturnType>(), &rhs)));
+    }
+    else if constexpr (!std::is_same<U, ReturnType>::value) 
+    {
+        static_assert(std::is_same<ReturnType, T>::value, "internal bug: lhs type is not the same as the computed common type");
+		ReturnType casted_rhs = static_cast<ReturnType>(rhs);
```

Comment:
Do not use tab to indent, use space. As you can see, using tabs works on your computer but fails to indent correctly here on github (if you are using an IDE, you can usually find an option to change this behavior).